Sources/MessagingInApp/Gist/Managers/MessageManager.swift

```diff
@@ -99,9 +101,37 @@ extension MessageManager: EngineWebDelegate {
         }
     }
 
+    // Function to check if a message has been previously clicked tracked
+    // If not, then track once & store in previouslyProcessedMessages
+    func trackInlineMessageIfNotPreviouslyTracked(action: String, name: String) {
+        let hasBeenTracked = previouslyProcessedMessages.contains { $0.id == currentMessage.id }
+        if !hasBeenTracked {
+            // Add the message to the list of previously processed messages
+            previouslyProcessedMessages.append(currentMessage)
+            trackTapAction(action: action, name: name)
+        }
+    }
+
+    func trackTapAction(action: String, name: String) {
+        // a close action does not count as a clicked action.
+        if action != "gist://close" {
+            if let deliveryId = getDeliveryId(from: currentMessage) {
+                eventBusHandler.postEvent(TrackInAppMetricEvent(deliveryID: deliveryId, event: InAppMetric.clicked.rawValue, params: ["actionName": name, "actionValue": action]))
+            }
+        }
+    }
+
     func tap(name: String, action: String, system: Bool) {
         Logger.instance.info(message: "Action triggered: \(action) with name: \(name)")
-        delegate?.action(message: currentMessage, currentRoute: currentRoute, action: action, name: name)
+        // This condition executes only for modal messages and not inline messages.
+        if currentMessage.isModalMessage {
+            delegate?.action(message: currentMessage, currentRoute: currentRoute, action: action, name: name)
+            trackTapAction(action: action, name: name)
+        } else {
+            // For inline messages, it prevents duplicate tracking and
+            // avoids making multiple event listener calls to delegate methods.
+            trackInlineMessageIfNotPreviouslyTracked(action: action, name: name)
+        }
```

Comment:
> I don't understand why this PR is only calling delegate.action() for modal messages?

It is because for inline messages we want to [call global event listener only when the delegate is not set](https://github.com/customerio/customerio-ios/blob/dd997c1a0da03fde4b267fc1284460fe651df609/Sources/MessagingInApp/Views/InAppMessageView.swift#L303) and if the delegate is set then we are required to call the custom callback.

To prevent multiple calls to the global listener, this check is required since we are catering to it in [InAppMessageView](https://github.com/customerio/customerio-ios/blob/dd997c1a0da03fde4b267fc1284460fe651df609/Sources/MessagingInApp/Views/InAppMessageView.swift#L301). I hope this make sense.


> With this updated description, I consider the prevent duplicate code in this class to be out-of-scope and it can be removed.

Right, I am updating the code based on the latest acceptance criteria as we talk.